### DISCUSSION TITLE
Primary key value for Identity key becomes invalid when using Polly

### DISCUSSION THREAD

**Discussion by balike888** [thepirat000/Audit.NET#674] [Q&A]
I've added Polly to my project to have some sort of fallback, but I've noticed this way the entry did not contain the freshly created Id from the databae (MS SQL) , instead an overflow/uninitialized value appeared:

{"Entry":{"Table":"X","Name":"XY","PrimaryKey":{"Id":-2147482645},"Action":"Insert","ColumnValues":...
"Entries":[{"Table":"X","Name":"XY","PrimaryKey":{"Id":-2147482645},"Action":"Insert","ColumnValues":{"Id":-2147482645,"Name":"1v"},"Valid":true}],"Result":0,"Success":false}}

According to this, the PrimaryKey value should be OK:

https://github.com/thepirat000/Audit.NET/discussions/646

```
var fallbackAudit = new SerilogDataProvider(_ => _
    .Logger(Log.Logger)
    .LogLevel(ev => ev.Environment.Exception != null ? Audit.Serilog.LogLevel.Fatal : Audit.Serilog.LogLevel.Info)
);

 Audit.Core.Configuration.Setup()
    .UsePolly(p => p
    .DataProvider(sqlDataProvider)
    .WithResilience(r => r
        .AddHedging(new()
        {
            ShouldHandle = new PredicateBuilder().Handle<SqlException>(),
            MaxHedgedAttempts = 2,
            Delay = TimeSpan.Zero,
            ActionGenerator = args => args.FallbackToDataProvider(fallbackAudit)
        }))
    );

```

Chaning back to the original

`Audit.Core.Configuration.Setup().UseCustomProvider(sqlDataProvider);`

version, the new Id appears fine.
Any suggestions, is this a known behavior?

**Comment by thepirat000** [comment]
I couldn't reproduce the issue.

I attempted to audit a DbContext by inserting a new entity with an auto-generated ID and configured a Polly strategy as follows:

```c#
var fallbackDataProvider = new FileDataProvider(c => c.Directory(@"c:\temp\logs"));

var sqlDataProvider = new SqlDataProvider(s => s
    .ConnectionString("data source=localhost;initial catalog=issue674;integrated security=true;Encrypt=False;")
    .TableName("Logs")
    .IdColumnName("Id")
    .CustomColumn("TableName", ev => ((AuditEventEntityFramework)ev).EntityFrameworkEvent.Entries[0].Table)
    .JsonColumnName("EventJson"));

Audit.Core.Configuration.Setup()
    .UsePolly(p => p
        .DataProvider(sqlDataProvider)
        .WithResilience(r => r
            .AddFallback(new()
            {
                ShouldHandle = new PredicateBuilder().Handle<Exception>(),
                FallbackAction = a => a.FallbackToDataProvider(fallbackDataProvider)
            })));
```

I see the correct primary key in both cases, (SQL or the file log)

**Reply by balike888** [reply]
Interesting! When I'm using your example, it works perfectly:

{"EntityFrameworkEvent":{"Database":"X","ConnectionId":"6d6eb8e8-7007-401b-b66e-87bd521298a5","ContextId":"919b3508-0049-4dd6-a02b-98f071e7394d:8","Entries":[{"Table":"Clients","Name":"PMClient","PrimaryKey":{"Id":14},"Action":"Insert","ColumnValues":{"Id":14,"Name":"ff"},"Valid":true},{"Table":"Accounts","Name":"PMAccount","PrimaryKey":{"Id":17},"Action":"Insert","ColumnValues":{"Id":17,"ClientId":14,"Enabled":true,"Name":"X"},"Valid":true}],"Result":2,"Success":true},"EventType":"WLPMSContext","Environment":...,"StartDate":"2024-07-04T07:40:22.2473409Z","EndDate":"2024-07-04T07:40:22.4067421Z","Duration":159}

However when I'm using Hedging, things go wrong:

                    .AddHedging(new()
                    {
                        ShouldHandle = new PredicateBuilder().Handle<SqlException>(),
                        MaxHedgedAttempts = 2,
                        Delay = TimeSpan.Zero,
                        ActionGenerator = args => args.FallbackToDataProvider(fallbackAudit)
                    })));

  {"EntityFrameworkEvent":{"Database":"X","ConnectionId":"e5b3bdf9-6352-4068-bd9e-e740a4af5ed5","ContextId":"97846a07-46fa-4118-8361-0a63ddf60dbc:56","Entries":[{"Table":"Clients","Name":"PMClient","PrimaryKey":{"Id":-**2147482647**},"Action":"Insert","ColumnValues":{"Id":-2147482647,"Name":"gnn11"},"Valid":true},{"Table":"Accounts","Name":"PMAccount","PrimaryKey":{"Id":-2147482647},"Action":"Insert","ColumnValues":{"Id":-**2147482647**,"ClientId":-2147482647,"Enabled":true,"Name":"X"},"Valid":true}],"Result":0,"Success":false},"EventType":"WLPMSContext","Environment":...,"StartDate":"2024-07-04T07:45:34.6097709Z","EndDate":"2024-07-04T07:45:34.6155424Z","Duration":6}

So I guess I've just used the wrong approach, thanks for help!
